### PR TITLE
Add test cases for nested role assignmnents

### PR DESCRIPTION
**Pull request by JPEasier**
# Description

Added a Pester test to check if the resource type in the nested roleAssignment bicep is the same as the module.

ToDo: add role assignments to all common test deployments

Issue Referenc: #3162

## Pipeline references

| Pipeline | Description |
| - | - |
| [Network - DNS Resolvers](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.dnsresolvers.yml) | failed pester test - wrong resource type |
| [KeyVault - Vaults](https://github.com/Azure/ResourceModules/actions/workflows/ms.keyvault.vaults.yml) | succeded pester test - correct resource type |

# Type of Change

Please delete options that are not relevant.

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: modules/Microsoft.Network/dnsResolvers/.test/common/deploy.test.bicep [hunk 1]
```diff
@@ -64,5 +64,14 @@ module testDeployment '../../deploy.bicep' = {
       Environment: 'Non-Prod'
       Role: 'DeploymentValidation'
     }
+    roleAssignments: [
+      {
+        roleDefinitionIdOrName: 'Reader'
+        description: 'Reader Role Assignment'
+        principalIds: [
+          'c99b4f9a-4268-4ab0-bd02-85d160b29a36' // carml-contributor-group
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/dnsResolvers/.test/common/deploy.test.bicep:72]: This should/must be the ID of an MSI that is deployed in the dependencies.

File: utilities/pipelines/staticValidation/module.tests.ps1 [hunk 2]
```diff
@@ -802,6 +802,38 @@ Describe 'Test file tests' -Tag 'TestTemplate' {
...
         }
     }
+    Context 'Role Assignmnet test' {
+
+        $roleAssignmentTestCases = @()
+
+        foreach ($moduleFolderPath in $moduleFolderPaths) {
+            if (Test-Path (Join-Path $moduleFolderPath '.bicep/nested_roleAssignments.bicep')) {
+                $nestedFilePath = Join-Path $moduleFolderPath '.bicep/nested_roleAssignments.bicep'
+                $nestedFileContent = Get-Content $nestedFilePath
```
Comment (AlexanderSehr) [inline utilities/pipelines/staticValidation/module.tests.ps1:812]:
```suggestion
           $rbacFilePath = Join-Path $moduleFolderPath '.bicep/nested_roleAssignments.bicep'
            if (Test-Path $rbacFilePath) {
                $nestedFileContent = Get-Content $rbacFilePath 
```

File: modules/Microsoft.Network/dnsResolvers/.test/common/deploy.test.bicep (hunk 1 above)
Comment (JPEasier) [reply modules/Microsoft.Network/dnsResolvers/.test/common/deploy.test.bicep:72]: True

File: utilities/pipelines/staticValidation/module.tests.ps1 [hunk 3]
```diff
@@ -802,6 +802,38 @@ Describe 'Test file tests' -Tag 'TestTemplate' {
...
+
+        $roleAssignmentTestCases = @()
+
+        foreach ($moduleFolderPath in $moduleFolderPaths) {
+           $rbacFilePath = Join-Path $moduleFolderPath '.bicep/nested_roleAssignments.bicep'
+            if (Test-Path $rbacFilePath) {
+                $nestedFileContent = Get-Content $rbacFilePath 
+
+                #$resourceTypeIdentifier = $moduleFolderPath.Replace('\', '/').Split('/modules/')[1]
+                $resourceTypeIdentifier = (Get-Content (Join-Path $moduleFolderPath 'readme.md')).split('`')[1].substring(1).Replace(']', '')
```
Comment (AlexanderSehr) [inline utilities/pipelines/staticValidation/module.tests.ps1:815]:
```suggestion
                $readMeContent = Get-Content (Join-Path $moduleFolderPath 'readme.md')
                $resourceTypeIdentifier = [regex]::Match($readMeContent[0], '^.*`\[([a-zA-Z\.\/]+)\]`$').Captures.Groups[1].Value
```

File: utilities/pipelines/staticValidation/module.tests.ps1 [hunk 4]
```diff
@@ -802,6 +802,38 @@ Describe 'Test file tests' -Tag 'TestTemplate' {
...
         }
     }
+    Context 'Role Assignmnet test' {
+
+        $roleAssignmentTestCases = @()
+
+        foreach ($moduleFolderPath in $moduleFolderPaths) {
+           $rbacFilePath = Join-Path $moduleFolderPath '.bicep/nested_roleAssignments.bicep'
+            if (Test-Path $rbacFilePath) {
+                $nestedFileContent = Get-Content $rbacFilePath
```
Comment (AlexanderSehr) [inline utilities/pipelines/staticValidation/module.tests.ps1:812]: I'd suggest to call it not `$nestedFileContent` but for example `$rbacFileContent` which would also match the `$rbacFilePath`

File: utilities/pipelines/staticValidation/module.tests.ps1 [hunk 5]
```diff
@@ -802,6 +802,38 @@ Describe 'Test file tests' -Tag 'TestTemplate' {
...
+        }
+
+        It '[<moduleFolderName>] nested role assignment deployment should contain the module resource Type as existing Resource.' -TestCases ($roleAssignmentTestCases | Where-Object { (Split-Path $_.nestedFilePath -Extension) -eq '.bicep' }) {
+
+            param(
+                [object[]] $nestedFileContent,
+                [string] $resourceTypeIdentifier
+            )
+
+            $hasExpectedResourceType = ($nestedFileContent | Out-String) -match $resourceTypeIdentifier
```
Comment (AlexanderSehr) [inline utilities/pipelines/staticValidation/module.tests.ps1:833]: While fine by itself, I would suggest to rather have a test that compares the used resource type with the one we expect as the `$resourceTypeIdentifier`. So the test would be like `$usedExistingResourceType | Should -Be $resourceTypeIdentifier`, which has the nice benefit, that a failing tests also tells you how you can fix the error.

To make this work however, you'd need to find the used resource type in the `$nestedFileContent`. This should be possible by searching for the line that matches `'resource ... existing = {'` and then extracting the used resource type.

This final step can be done using the snippet
```PowerShell
[regex]::Match($nestedFileContent[$existingResourceTypeRow], "^resource.+'([a-zA-Z\.\/]+)@.+' existing = {$").Captures.Groups[1].Value
```

**Pull request closed without merging**